cpp/src/arrow/compute/exec/asof_join_node.cc

```diff
@@ -137,18 +167,84 @@ struct MemoStore {
   }
 };
 
+// a specialized higher-performance variation of Hashing64 logic
```

Comment:
Can you explain the difference between this and the existing Hashing64 classes and why it is faster?

Reply:
The doc string is a bit misleading. The logic here is faster than the [`Hashing64` logic in `hash_join`](https://github.com/apache/arrow/blob/d880d7517a33f2ac8ff259cad711bc210fd570c5/cpp/src/arrow/compute/exec/hash_join_node.cc#L1164-L1178), because the latter recreates/recomputes certain objects on every batch while the former avoids that.